resources/netregexes.ts

```diff
@@ -36,32 +38,87 @@ const defaultParams = <
   T extends LogDefinitionTypes,
   V extends LogDefinitionVersions,
 >(type: T, version: V, include?: string[]): Partial<ParseHelperFields<T>> => {
-  include ??= Object.keys(logDefinitionsVersions[version][type].fields);
-  const params: { [index: number]: { field: string; value?: string; optional: boolean } } = {};
-  const firstOptionalField = logDefinitionsVersions[version][type].firstOptionalField;
+  const logType = logDefinitionsVersions[version][type];
+  if (include === undefined) {
+    include = Object.keys(logType.fields);
+    if ('repeatingFields' in logType) {
+      include.push(logType.repeatingFields.label);
+    }
+  }
 
-  for (const [prop, index] of Object.entries(logDefinitionsVersions[version][type].fields)) {
+  const params: {
+    [index: number]: {
+      field: string;
+      value?: string;
+      optional: boolean;
+      repeating?: boolean;
+      repeatingKeys?: string[];
+      sortFn?: (L: unknown, R: unknown) => number;
+    };
+  } = {};
+  const firstOptionalField = logType.firstOptionalField;
+
+  for (const [prop, index] of Object.entries(logType.fields)) {
     if (!include.includes(prop))
       continue;
-    const param: { field: string; value?: string; optional: boolean } = {
+    const param: { field: string; value?: string; optional: boolean; repeating?: boolean } = {
       field: prop,
       optional: firstOptionalField !== undefined && index >= firstOptionalField,
     };
     if (prop === 'type')
-      param.value = logDefinitionsVersions[version][type].type;
+      param.value = logType.type;
 
     params[index] = param;
   }
 
+  if ('repeatingFields' in logType && include.includes(logType.repeatingFields.label)) {
+    params[logType.repeatingFields.startingIndex] = {
+      field: logType.repeatingFields.label,
+      optional: firstOptionalField !== undefined &&
+        logType.repeatingFields.startingIndex >= firstOptionalField,
+      repeating: true,
+      repeatingKeys: [...logType.repeatingFields.names],
+      sortFn: logType.repeatingFields.sortFn,
+    };
+  }
+
   return params as unknown as Partial<ParseHelperFields<T>>;
 };
 
-type ParseHelperType<T extends LogDefinitionTypes> =
+type RepeatingFieldsMap<
+  TBase extends LogDefinitionTypes,
+  TKey extends RepeatingFieldsTypes = TBase extends RepeatingFieldsTypes ? TBase : never,
+> = {
+  // This syntax is a bit weird to make typescript compiler happy
+  [
+    label in RepeatingFieldsDefintions[TKey]['repeatingFields'][
+      'label'
+    ] as `${RepeatingFieldsDefintions[TKey]['repeatingFields']['label']}`
+  ]?: {
+    [name in RepeatingFieldsDefintions[TKey]['repeatingFields']['names'][number]]:
+      | string
+      | string[];
+  }[];
+};
+
+type RepeatingFieldsMapType<T extends LogDefinitionTypes> = T extends RepeatingFieldsTypes
+  ? RepeatingFieldsMap<T>
+  : Record<string, unknown>;
```

Comment:
This is a combined reply for both this comment and https://github.com/quisquous/cactbot/pull/5379#discussion_r1193194445.

I think you're maybe misunderstanding the underlying additional type here, which is what's causing the confusion. Consider the following more explicit definitions:

```ts
type RepeatingFieldsMapSubtype<
  TBase extends LogDefinitionTypes,
  TKey extends RepeatingFieldsTypes = TBase extends RepeatingFieldsTypes ? TBase : never,
> = TBase extends LogDefinitionTypes ? {
  [name in RepeatingFieldsDefinitions[TKey]['repeatingFields']['names'][number]]:
    | string
    | string[];
} : never;

type RepeatingFieldsMap<
  TBase extends LogDefinitionTypes,
  TKey extends RepeatingFieldsTypes = TBase extends RepeatingFieldsTypes ? TBase : never,
> = TBase extends LogDefinitionTypes ? {
  [
    label in RepeatingFieldsDefinitions[TKey]['repeatingFields']['label']
  ]?: RepeatingFieldsMapSubtype<TKey>[];
} : never;

// ...

    let fieldValue:
      string | string[] | RepeatingFieldsMapSubtype<T>[] | undefined = params[fieldName];

    if (fields[keyStr]?.repeating) {
      if (Array.isArray(fieldValue)) {
        let repeatingArray: RepeatingFieldsMapSubtype<T>[] =
          fieldValue as unknown as RepeatingFieldsMapSubtype<T>[];
```

This requires an even more ugly cast, because without any cast:

![image](https://github.com/quisquous/cactbot/assets/6119598/d54c72b0-de40-489e-8887-397e4bf1b26c)
![image](https://github.com/quisquous/cactbot/assets/6119598/a886fbd8-d8bc-40f5-be2b-fe4a8fa82d2c)
![image](https://github.com/quisquous/cactbot/assets/6119598/d95f7630-c155-4586-a7cb-11731cc333e5)

As for why `RepeatingFieldsMapSubtype<T>[]` is not in the typedef, I have no clue. It won't recognize any array type added to the `fieldValue` typing beyond `string[]`.

Reply:
I have refactored the logic for these types in a way that hopefully makes more sense.

Reply:
Thanks! I appreciate all the changes.